File: +fs/+lpew/+v2/ksInterp.m
function [Kt1, Kt2, Kn1, Kn2] = ksInterp(FS, T_all, Q_corner)
%FS.LPEW.V2.KSINTERP  Batched permeability-tensor projections (replaces ferncodes_Ks_Interp_LPEW2).
%
%   [Kt1, Kt2, Kn1, Kn2] = fs.lpew.v2.ksInterp(FS, T, Qcorner)
%
%   Inputs:
%     T          [nAllNeighbors x 3]  midpoints (from fs.lpew.OPT)
%     Qcorner    [nCorners       x 3]  owning node coord per corner (from fs.lpew.OPT)
%
%   Preconditions:
%     FS.perm.kmap             — [nElems x 5+]  legacy kmap layout (col 2-5 = K11/K12/K21/K22)
%     FS.mesh.elem             — [nElems x 5+]  col 5 = material id
%     FS.csr.tCurrent/tNext    — per-corner shift indices (from buildCornerShifts)
%     FS.csr.cornerElem        — per-corner element id
%     FS.cfg.phasekey          — 1 (single-phase → L22=1) OR 2 (two-phase, requires Sw)
%
%   Outputs (per corner, one row per corner):
%     Kt1  [nCorners x 2]   tangential projections, column 1 for i=1 (T(k)),
%                           column 2 for i=2 (T(k+1) with interior wrap)
%     Kt2  [nCorners x 1]   tangential projection along T(k+1)-T(k)
%     Kn1  [nCorners x 2]   normal projections
%     Kn2  [nCorners x 1]   normal projection along T(k+1)-T(k)
%
%   Math: R = 90° rotation in 2D (embedded in 3D). For vector v = [vx, vy, 0]':
%     Rv = [vy, -vx, 0]'
%     <Rv, K*Rv> = K11*vy² - (K12+K21)*vx*vy + K22*vx²          → normal projection
%     <Rv, K*v>  = (K11-K22)*vx*vy + K12*vy² - K21*vx²          → tangential projection
%
%   For two-phase (phasekey=2), Kn2/Kt2 use L22-scaled tensor via twophasevar. Not
%   yet vectorized — falls back to error until phasekey=2 use case appears in tests.

    coord      = FS.mesh.coord;                                                    %#ok<NASGU>  % kept for future use
    elem       = FS.mesh.elem;
    kmap       = FS.perm.kmap;
    cornerElem = FS.csr.cornerElem;
    tCurrent   = FS.csr.tCurrent;
    tNext      = FS.csr.tNext;

    if isfield(FS.cfg, 'phasekey') && FS.cfg.phasekey == 2
        error('fs.lpew.v2.ksInterp:TwoPhaseNotVectorized', ...
              'Two-phase (phasekey=2) mobility scaling not yet vectorized. Track: PR-F.');
    end

    % Per-corner tensor (K11, K12, K21, K22) from element material id
    matIds = elem(cornerElem, 5);          % [nCorners x 1]
    % Some elements may have matId=0 (unset). Guard by clamping to 1 (matches
    % legacy behavior — MATLAB would error on 0 index, but sensible legacy meshes
    % have matId>=1). If any 0s exist, legacy would have crashed too.
    K11 = kmap(matIds, 2);
    K12 = kmap(matIds, 3);
    K21 = kmap(matIds, 4);
    K22 = kmap(matIds, 5);

    % ── Column i=1: use T(k) (i.e. tCurrent) ─────────────────────────
    v_k  = T_all(tCurrent, :) - Q_corner;  % [nCorners x 3]
    vx = v_k(:, 1); vy = v_k(:, 2);
    len2_k = vx.*vx + vy.*vy;
    Kn1_c1 = (K11.*vy.*vy - (K12 + K21).*vx.*vy + K22.*vx.*vx) ./ len2_k;
    Kt1_c1 = ((K11 - K22).*vx.*vy + K12.*vy.*vy - K21.*vx.*vx) ./ len2_k;

    % ── Column i=2: use T(k+1) (i.e. tNext — wraps for interior last corner) ─
    v_k1 = T_all(tNext, :) - Q_corner;
    vx = v_k1(:, 1); vy = v_k1(:, 2);
    len2_k1 = vx.*vx + vy.*vy;
    Kn1_c2 = (K11.*vy.*vy - (K12 + K21).*vx.*vy + K22.*vx.*vx) ./ len2_k1;
    Kt1_c2 = ((K11 - K22).*vx.*vy + K12.*vy.*vy - K21.*vx.*vx) ./ len2_k1;

    Kn1 = [Kn1_c1, Kn1_c2];
    Kt1 = [Kt1_c1, Kt1_c2];

    % ── Kn2, Kt2: use v = T(k+1) - T(k) with L22-scaled tensor ───────
    % Single-phase: L22=1 → K1 = K. Two-phase would multiply K by L22 per corner.
    v_diff = T_all(tNext, :) - T_all(tCurrent, :);
    vx = v_diff(:, 1); vy = v_diff(:, 2);
    len2_diff = vx.*vx + vy.*vy;
    Kn2 = (K11.*vy.*vy - (K12 + K21).*vx.*vy + K22.*vx.*vx) ./ len2_diff;
    Kt2 = ((K11 - K22).*vx.*vy + K12.*vy.*vy - K21.*vx.*vx) ./ len2_diff;
end

File: fs/lpew/v2/ksInterp.py
"""Permeability projections translated from ``+fs/+lpew/+v2/ksInterp.m``."""

from typing import Any, Dict, Tuple

import numpy as np


def ksInterp(
    FS: Dict[str, Any],
    T_all: np.ndarray,
    Q_corner: np.ndarray,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """Return tangential and normal permeability projections per corner."""
    if FS.get("cfg", {}).get("phasekey") == 2:
        raise NotImplementedError(
            "Two-phase mobility scaling has not been vectorized for ksInterp"
        )

    mesh = FS["mesh"]
    csr = FS["csr"]
    kmap = np.asarray(FS["perm"]["kmap"], dtype=float)
    elem = np.asarray(mesh["elem"])
    corner_elem = np.asarray(csr["cornerElem"], dtype=int)
    t_current = np.asarray(csr["tCurrent"], dtype=int)
    t_next = np.asarray(csr["tNext"], dtype=int)

    material_ids = np.asarray(elem[corner_elem, 4], dtype=int)
    material_rows = _material_rows(kmap, material_ids)

    k11 = kmap[material_rows, 1]
    k12 = kmap[material_rows, 2]
    k21 = kmap[material_rows, 3]
    k22 = kmap[material_rows, 4]

    T_all = np.asarray(T_all, dtype=float)
    Q_corner = np.asarray(Q_corner, dtype=float)

    kt1_col1, kn1_col1 = _project(
        T_all[t_current] - Q_corner, k11, k12, k21, k22
    )
    kt1_col2, kn1_col2 = _project(
        T_all[t_next] - Q_corner, k11, k12, k21, k22
    )
    kt2, kn2 = _project(
        T_all[t_next] - T_all[t_current], k11, k12, k21, k22
    )

    kt1 = np.column_stack((kt1_col1, kt1_col2))
    kn1 = np.column_stack((kn1_col1, kn1_col2))
    return kt1, kt2, kn1, kn2


def _material_rows(kmap: np.ndarray, material_ids: np.ndarray) -> np.ndarray:
    if kmap.ndim != 2 or kmap.shape[1] < 5:
        raise ValueError("FS.perm.kmap must contain material id and K11/K12/K21/K22")

    row_by_id = {int(material_id): row for row, material_id in enumerate(kmap[:, 0])}
    missing = sorted(set(material_ids) - row_by_id.keys())
    if missing:
        raise ValueError(f"Material ids missing from FS.perm.kmap: {missing}")
    return np.fromiter(
        (row_by_id[material_id] for material_id in material_ids),
        dtype=int,
        count=material_ids.size,
    )


def _project(
    vectors: np.ndarray,
    k11: np.ndarray,
    k12: np.ndarray,
    k21: np.ndarray,
    k22: np.ndarray,
) -> Tuple[np.ndarray, np.ndarray]:
    vx = vectors[:, 0]
    vy = vectors[:, 1]
    length_squared = vx * vx + vy * vy
    if np.any(length_squared == 0):
        raise ValueError("Cannot project permeability along a zero-length vector")

    normal = (
        k11 * vy * vy
        - (k12 + k21) * vx * vy
        + k22 * vx * vx
    ) / length_squared
    tangential = (
        (k11 - k22) * vx * vy
        + k12 * vy * vy
        - k21 * vx * vx
    ) / length_squared
    return tangential, normal
